Add FOREACH to cost planner

henriknyman opened the pull request.

henriknyman committed:
Add FOREACH to cost planner

henriknyman committed:
Fix so node variables can be used inside foreach

Add the foreach variable to query graph pattern node if it has type node.
This will make create patterns be planned correctly with what has already
been matched.

henriknyman committed:
Enable all foreach tests to run with cost planner

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/ast/convert/plannerQuery/ClauseConverters.scala:
```diff
@@ -381,6 +384,28 @@ object ClauseConverters {
+      expression = clause.expression,
+      innerUpdates = innerPlannerQuery)
+
+    builder.amendQueryGraph(_.addMutatingPatterns(foreachPattern))
```
foreach is not necessarily a mutating thing

Mats-SX replied:
Not in itself, but its purpose is definitely tied to updates, since it allows only update clauses in its body.

pontusmelke replied:
oh I didn't know that.

Mats-SX replied:
Don't feel bad; I didn't know that either until we did this work.

Mats-SX commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/ForeachPipe.scala:
```diff
@@ -0,0 +1,59 @@
+        val values = makeTraversable(expression(outerContext)(state))
+        values.foreach { v =>
+          val innerState = state.withInitialContext(outerContext.newWith1(variable, v))
+          // do we need to exhaust this iterator?
```
Well, do we? Acceptance tests should show if we need to or not; we should get rid of this comment.

systay replied:
This is an iterator purely consumed for it's side effects, right? So something needs to exhaust it. What else would?

I'm pretty sure we need to exhaust it.

Mats-SX replied:
You're right, and the reason is the merge read, which may increase cardinality. In all other cases, there's nothing to exhaust (only one row per `FOREACH` iteration).

henriknyman committed:
Fix planning of foreach inner updates

henriknyman committed:
Support eagerness with foreach

- Conservative eagerness rule for foreach (always assume conflict)
- Split out foreach into its own planner query
- Added eagerness acceptance tests
- Added unit test
- Fixed bug in createNodeOverlap

Eagerness for nested foreach statements still remains to be done

henriknyman committed:
Fix eagerness with nested foreach

- Change planning of merge to be more aligned with other update clauses
 (combine read/write parts)
 - Isolate planning of eagerness from other planning segments
 - Fix eagerness bug where pattern relationships from optional matches were
  not being considered
 - Proper passing of arguments through horizons/projections
 - Remove unused arguments from merge query graphs
 - Unnest foreach from inside apply

systay commented on community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala:
```diff
@@ -1673,7 +1673,7 @@ return b
     val first = eengine.execute(query).toList
     val second = eengine.execute(query).toList
-    val check = executeWithAllPlannersAndRuntimes("MATCH (f:Folder) RETURN f.name").toSet
+    val check = updateWithBothPlanners("MATCH (f:Folder) RETURN f.name").toSet
```
Why?

Mats-SX replied:
Yeah, this looks wrong.

systay commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/plans/Foreach.scala:
```diff
@@ -0,0 +1,32 @@
+import org.neo4j.cypher.internal.compiler.v3_0.planner.{CardinalityEstimation, PlannerQuery}
+import org.neo4j.cypher.internal.frontend.v3_0.ast.Expression
+
+case class Foreach(left: LogicalPlan, right: LogicalPlan, variable: String, expression: Expression)(val solved: PlannerQuery with CardinalityEstimation)
```
If I understand it correctly, this is another one of those plans that feed something to an Argument and then does something with an inner plan. I would argue that we should call this logical plan operator `ForeachApply`. WDYT?

Mats-SX committed:
Settle implementor uncertainty

Mats-SX committed:
Improve test for repeated `MERGE` and `FOREACH`

Mats-SX committed:
Rename foreach operator to make category explicit
